File: 1094-car-pooling/1094-car-pooling.cpp
class Solution {
public:
    
static bool sortcol(const vector<int>& v1, const vector<int>& v2)
{
    return v1[1] < v2[1];
}
    
    bool carPooling(vector<vector<int>>& trips, int capacity) 
    {
        sort(trips.begin(), trips.end(), sortcol);
        
       priority_queue <pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
        
        // for(int i=0; i<trips.size(); i++)
        // {
        //     for(int j=0; j<trips[i].size(); j++)
        //     {
        //         cout<<trips[i][j]<<" ";
        //     }
        //     cout<<endl;
        // }
        
        for(int i=0; i<trips.size(); i++)
        {
            while(!pq.empty() && trips[i][1] >= pq.top().first)
            {
                capacity += pq.top().second;
                pq.pop();
            }
            
            if(trips[i][0] <= capacity)
            {
                capacity -= trips[i][0];
                pair<int,int> pr;
                pr.first = trips[i][2];
                pr.second = trips[i][0];
                
                pq.push(pr);
            }
            
            else
            {
                return false;
            }
        }
        
        return true;
    }
};

File: 1094-car-pooling/1094. Car Pooling.java
class Solution {
    public boolean carPooling(int[][] trips, int capacity) 
    {
        int[] preSum = new int[1001];

        int n = trips.length;

        for(int i=0; i<n; i++)
        {
            int people = trips[i][0];
            int from = trips[i][1];
            int to = trips[i][2];

            preSum[from] += people;
            preSum[to] -= people;
        }

        for(int i=1; i<1001; i++)preSum[i] += preSum[i-1];

        for(Integer it : preSum)if(it > capacity)return false;
        return true;
    }
}